### DISCUSSION TITLE
JWT role claim

### DISCUSSION THREAD

**Discussion by tbmale** [prest/prest#363] [API Server] [answered]
Hello,

  I couldn't find any reference in the docs nor the issues about using the JWT claims.
Can I use the role claim from an JWT token as the pg role for the session ? (à la postgrest)

Thanks

**Comment by pyramation** [comment]
this library looks awesome! I really want to use prest, however, I'm in the same boat as @tbmale .... just looking for some docs on how to use this or what is it's default behavior/expectations

Are there any examples out there for integrating a JWT system? 

Here is how I've done it and would like to:

### issuing and storing tokens/sessions

The way I typically do it, is I've actually implemented a table with my own tokens, and a function for validating and issuing tokens to users. The table is private, and never exposed to public. The function will check against a users email/password to issue a token. 

### authenticating requests

Then when client sends requests, token is pulled off to be authenticated, before every request, and then sets some settings using `set_config`, which changes `role` and gives some meta-data that my policies use for RLS. This is set at the beginning of the request in a transaction, so the information goes away after transaction is commit:

```sql
select set_config('role', 'app_user', true);
select set_config('jwt.claims.role_id', '681f005b-f897-467f-86de-8f4e59bfb12f', true);
```

### write SQL using this information in procedures and policies

Now in using RLS and policies, you can use this information:

```sql
CREATE POLICY can_do_something ON some_table
  FOR UPDATE
  USING (
    current_setting('jwt.claims.role_id')::uuid = sender_id
  );
```

**Comment by pyramation** [comment]
@avelino @joelmdesouza

**Comment by avelino** [comment]
We are working on the implementation of standard authentication, can be followed here https://github.com/prest/prest/issues/276

**Comment by avelino** [comment]
cc @tiaguinho

**Comment by avelino** [comment, marked as answer]
implementation of authentication endpoint completed, we will be grateful for your feedback
view more: https://docs.postgres.rest/query-statements/#auth-auth---post

**Reply by nicowilliams** [reply]
> implementation of authentication endpoint completed, we will be grateful for your feedback view more: https://docs.postgres.rest/query-statements/#auth-auth---post

That link is dead, and it does not answer the question anyways.  I could not find anything in the source code to `prest` that makes the user-agent's/client's JWT's claims available to the SQL code (views, functions, triggers) in any way.  I'd have expected that the claims would be made available via calling the PG function `set_config()` on the connection and session used for processing a given HTTP/REST request.  But there's no sign of this.  I opened a new discussion, #924 about this.
